modal/_functions.py

```diff
@@ -970,20 +970,7 @@ def _bind_parameters(
         parent = self
 
         async def _load(param_bound_func: _Function, resolver: Resolver, existing_object_id: Optional[str]):
-            try:
-                identity = f"{parent.info.function_name} class service function"
-            except Exception:
-                # Can't always look up the function name that way, so fall back to generic message
-                identity = "class service function for a parametrized class"
-            if not parent.is_hydrated:
-                if parent.app._running_app is None:
-                    reason = ", because the App it is defined on is not running"
-                else:
-                    reason = ""
-                raise ExecutionError(
-                    f"The {identity} has not been hydrated with the metadata it needs to run on Modal{reason}."
-                )
-
+            await parent.hydrate()
```

Comment:
Very satisfying patch!

Yeah I think this is mostly old legacy stuff from pre-lazily-hydratable classes.

There is still one code path where classes aren't lazily hydratable, which we should make sure has ok error message: when someone tries to use an in-project defined class without having run/deployed the app, e.g.:

```py

@app.cls()
class A:
     ...

if __name__ == "__main__":
     # no app.run() session etc. makes this usage illegal and it shoud have a good error message,
     # similar to the old `App it is defined on is not running`
     A().keep_warm(...)  # not inside a 

```

However this error message would probably belong somewhere in the *base* function hydration step (i.e. within `parent.hydrate()` here)

Reply:
Yep this still triggers from the underlying `.hydrate` call. The error message is more generic:

```
modal.exception.ExecutionError: Function has not been hydrated with the metadata it needs to run on Modal, because the App it is defined on is not running.
```

rather than the more custom

```
modal.exception.ExecutionError: The T.* class service function has not been hydrated with the metadata it needs to run on Modal, because the App it is defined on is not running.
```

I think that's acceptable since "class service function" isn't really a public-facing concept anyway.